### PR TITLE
Add POST and PUT examples in assets API doc

### PR DESCRIPTION
**Pull request by hillaryfraley**
## Description
Add POST /assets and PUT and DELETE /assets/:asset examples.

Epic #1452

## Review Instructions
I replaced all the code examples with examples that have `"filters": null,` because when I try to run POST or PUT with the current examples that include filters:
```
{
    "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
    "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
    "filters": {
      "entity.system.os": "linux",
      "entity.system.arch": "amd64"
    },
    "headers": null,
    "metadata": {
      "name": "sensu-slack-handler",
      "namespace": "default"
    }
 }
```

...I receive this error:
`{"message":"json: cannot unmarshal object into Go struct field Asset.filters of type []string","code":1}`

### CODE REVIEW

File: content/sensu-go/5.16/api/assets.md [hunk 1]
```diff
@@ -28,22 +28,32 @@ The `/assets` API endpoint provides HTTP GET access to [asset][1] data.
 The following example demonstrates a request to the `/assets` API endpoint, resulting in a JSON array that contains [asset definitions][1].
 
 {{< highlight shell >}}
-curl http://127.0.0.1:8080/api/core/v2/namespaces/default/assets -H "Authorization: Bearer $SENSU_TOKEN"
+curl -X GET http://127.0.0.1:8080/api/core/v2/namespaces/default/assets
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: requires ' \ ' at the end of the line, probably should move 'curl -X GET' to it's own line for consistency

File: content/sensu-go/5.16/api/assets.md [hunk 2]
```diff
@@ -125,21 +155,20 @@ The `/assets/:asset` API endpoint provides HTTP GET access to [asset data][1] fo
 In the following example, querying the `/assets/:asset` API endpoint returns a JSON map that contains the requested [`:asset` definition][1] (in this example, for the `:asset` named `check_script`).
 
 {{< highlight shell >}}
-curl http://127.0.0.1:8080/api/core/v2/namespaces/default/assets/sensu-slack-handler -H "Authorization: Bearer $SENSU_TOKEN"
+curl -X GET http://127.0.0.1:8080/api/core/v2/namespaces/default/assets/sensu-slack-handler
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: requires ' \ ' at the end of the line, probably should move 'curl -X GET' to it's own line for consistency

**Review by Blue0ctober — CHANGES_REQUESTED** [review]
See comments for recommended changes

File: content/sensu-go/5.16/api/assets.md (hunk 1 above)
Comment (hillaryfraley) [inline content/sensu-go/5.16/api/assets.md]:
```suggestion
curl -X GET \
http://127.0.0.1:8080/api/core/v2/namespaces/default/assets \
```

File: content/sensu-go/5.16/api/assets.md (hunk 2 above)
Comment (hillaryfraley) [inline content/sensu-go/5.16/api/assets.md]:
```suggestion
curl -X GET \
http://127.0.0.1:8080/api/core/v2/namespaces/default/assets/sensu-slack-handler \
```

**Comment by hillaryfraley** [comment]
Thanks @Blue0ctober I fixed the two curl examples as you noted.

Do you know who I could ask about the filters issue in the original PR description? Or can you check if you get the same result when you run the POST and PUT calls with filters as noted in the PR description?

File: content/sensu-go/5.16/api/assets.md [hunk 3]
```diff
@@ -28,22 +28,33 @@ The `/assets` API endpoint provides HTTP GET access to [asset][1] data.
...
+http://127.0.0.1:8080/api/core/v2/namespaces/default/assets \
+-H "Authorization: Bearer $SENSU_TOKEN" \
 
 HTTP/1.1 200 OK
 [
+  {
+    "url": "https://github.com/sensu/sensu-influxdb-handler/releases/download/3.1.2/sensu-influxdb-handler_3.1.2_linux_amd64.tar.gz",
+    "sha512": "612c6ff9928841090c4d23bf20aaf7558e4eed8977a848cf9e2899bb13a13e7540bac2b63e324f39d9b1257bb479676bc155b24e21bf93c722b812b0f15cb3bd",
+    "filters": null,
+    "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md:44]: do we need to include filters and builds at all if they're going to be `null`?  might be better to just remove them instead?

File: content/sensu-go/5.16/api/assets.md [hunk 4]
```diff
@@ -28,22 +28,33 @@ The `/assets` API endpoint provides HTTP GET access to [asset][1] data.
 The following example demonstrates a request to the `/assets` API endpoint, resulting in a JSON array that contains [asset definitions][1].
 
 {{< highlight shell >}}
-curl http://127.0.0.1:8080/api/core/v2/namespaces/default/assets -H "Authorization: Bearer $SENSU_TOKEN"
+curl -X GET \
+http://127.0.0.1:8080/api/core/v2/namespaces/default/assets \
+-H "Authorization: Bearer $SENSU_TOKEN" \
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: remove the last \ on the last multi line command

File: content/sensu-go/5.16/api/assets.md [hunk 5]
```diff
@@ -28,22 +28,33 @@ The `/assets` API endpoint provides HTTP GET access to [asset][1] data.
...
   {
     "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
     "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
-    "filters": {
-      "entity.system.os": "linux",
-      "entity.system.arch": "amd64"
-    },
-    "headers": null,
+    "filters": null,
+    "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md:61]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 6]
```diff
@@ -62,24 +73,24 @@ output         | {{< highlight shell >}}
   {
     "url": "https://github.com/sensu/sensu-influxdb-handler/releases/download/3.1.2/sensu-influxdb-handler_3.1.2_linux_amd64.tar.gz",
     "sha512": "612c6ff9928841090c4d23bf20aaf7558e4eed8977a848cf9e2899bb13a13e7540bac2b63e324f39d9b1257bb479676bc155b24e21bf93c722b812b0f15cb3bd",
-    "filters": {
-      "entity.system.os": "linux",
-      "entity.system.arch": "amd64"
-    },
-    "headers": null,
+    "filters": null,
+    "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 7]
```diff
@@ -62,24 +73,24 @@ output         | {{< highlight shell >}}
...
     "metadata": {
       "name": "sensu-influxdb-handler",
       "namespace": "default"
-    }
+    },
+    "headers": null
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 8]
```diff
@@ -62,24 +73,24 @@ output         | {{< highlight shell >}}
...
   },
   {
     "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
     "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
     "filters": null,
+    "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md:109]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 9]
```diff
@@ -88,6 +99,31 @@ output         | {{< highlight shell >}}
...
+-d '{
+  "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
+  "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
+  "filters": null,
+  "builds": null,
+  "metadata": {
+    "name": "sensu-slack-handler",
+    "namespace": "default"
+  },
+  "headers": null
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 10]
```diff
@@ -96,23 +132,18 @@ description     | Creates a Sensu asset.
...
-    "entity.system.arch": "amd64"
-  },
-  "headers": {
-    "Authorization": "Bearer $TOKEN",
-    "X-Forwarded-For": "client1, proxy1, proxy2"
-  },
+  "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
+  "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
+  "filters": null,
+  "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 11]
```diff
@@ -96,23 +132,18 @@ description     | Creates a Sensu asset.
...
   "metadata": {
-    "name": "my-secure-asset",
+    "name": "sensu-slack-handler",
     "namespace": "default"
-  }
+  },
+  "headers": null
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 12]
```diff
@@ -125,21 +156,21 @@ The `/assets/:asset` API endpoint provides HTTP GET access to [asset data][1] fo
...
 {
   "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
   "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
-  "filters": {
-    "entity.system.os": "linux",
-    "entity.system.arch": "amd64"
-  },
-  "headers": null,
+  "filters": null,
+  "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 13]
```diff
@@ -125,21 +156,21 @@ The `/assets/:asset` API endpoint provides HTTP GET access to [asset data][1] fo
...
   "metadata": {
     "name": "sensu-slack-handler",
     "namespace": "default"
-  }
+  },
+  "headers": null
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 14]
```diff
@@ -148,52 +179,69 @@ HTTP/1.1 200 OK
...
-    "entity.system.arch": "amd64"
-  },
-  "headers": {
-    "Authorization": "Bearer $TOKEN",
-    "X-Forwarded-For": "client1, proxy1, proxy2"
-  },
+  "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
+  "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
+  "filters": null,
+  "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 15]
```diff
@@ -148,52 +179,69 @@ HTTP/1.1 200 OK
...
   "metadata": {
-    "name": "my-secure-asset",
+    "name": "sensu-slack-handler",
     "namespace": "default"
-  }
+  },
+  "headers": null
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 16]
```diff
@@ -148,52 +179,69 @@ HTTP/1.1 200 OK
...
+
+{{< highlight shell >}}
+curl -X PUT \
+-H "Authorization: Bearer $SENSU_TOKEN" \
+-H 'Content-Type: application/json' \
+-d '{
+  "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
+  "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
+  "filters": null,
+  "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 17]
```diff
@@ -148,52 +179,69 @@ HTTP/1.1 200 OK
...
+  "metadata": {
+    "name": "sensu-slack-handler",
+    "namespace": "default"
+  },
+  "headers": null
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 18]
```diff
@@ -148,52 +179,69 @@ HTTP/1.1 200 OK
...
-    "entity.system.arch": "amd64"
-  },
-  "headers": {
-    "Authorization": "Bearer $TOKEN",
-    "X-Forwarded-For": "client1, proxy1, proxy2"
-  },
+  "url": "https://github.com/sensu/sensu-slack-handler/releases/download/1.0.3/sensu-slack-handler_1.0.3_linux_amd64.tar.gz",
+  "sha512": "68720865127fbc7c2fe16ca4d7bbf2a187a2df703f4b4acae1c93e8a66556e9079e1270521999b5871473e6c851f51b34097c54fdb8d18eedb7064df9019adc8",
+  "filters": null,
+  "builds": null,
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

File: content/sensu-go/5.16/api/assets.md [hunk 19]
```diff
@@ -148,52 +179,69 @@ HTTP/1.1 200 OK
...
   "metadata": {
-    "name": "my-secure-asset",
+    "name": "sensu-slack-handler",
     "namespace": "default"
-  }
+  },
+  "headers": null
```
Comment (Blue0ctober) [inline content/sensu-go/5.16/api/assets.md]: same as above re: null

**Review by Blue0ctober — CHANGES_REQUESTED** [review]
see comments for changes

File: content/sensu-go/5.16/api/assets.md (hunk 3 above)
Comment (hillaryfraley) [reply content/sensu-go/5.16/api/assets.md:44]: Another question is whether you can create assets with filters (i.e., can you add filters later if you exclude them when you initially create an asset?).

I could not get the calls to succeed with a filter specified. The calls to create an asset do work with filters:null or leaving the filters array out altogether.

**Comment by cwjohnston** [comment]
> {"message":"json: cannot unmarshal object into Go struct field Asset.filters of type []string","code":1}

I believe this is telling us that the `filters` attribute is expected to have a value of type list (array) containing string values.

e.g. `{ "filters": [ "event.check.occurrences == 1" ]}`

File: content/sensu-go/5.16/api/assets.md (hunk 3 above)
Comment (hillaryfraley) [reply content/sensu-go/5.16/api/assets.md:44]: Cameron's filters format suggestion worked, so this is fixed now!

**Pull request merged**